pandas/hashtable.pyx

```diff
@@ -1026,24 +1026,40 @@ def mode_int64(int64_t[:] values):
 
 @cython.wraparound(False)
 @cython.boundscheck(False)
-def duplicated_int64(ndarray[int64_t, ndim=1] values, int take_last):
+def duplicated_int64(ndarray[int64_t, ndim=1] values, object keep='first'):
     cdef:
         int ret = 0
         Py_ssize_t i, n = len(values)
         kh_int64_t * table = kh_init_int64()
         ndarray[uint8_t, ndim=1, cast=True] out = np.empty(n, dtype='bool')
 
+        # for keep=all to preserve occurrence loc
+        dict seen = dict()
+
     kh_resize_int64(table, min(n, _SIZE_HINT_LIMIT))
 
-    with nogil:
-        if take_last:
+    if keep == 'last':
+        with nogil:
             for i from n > i >=0:
                 kh_put_int64(table, values[i], &ret)
                 out[i] = ret == 0
-        else:
+    elif keep == 'first':
+        with nogil:
             for i from 0 <= i < n:
                 kh_put_int64(table, values[i], &ret)
                 out[i] = ret == 0
+    elif keep is False:
+        # unable to release gil to use dict
```

Comment:
sure. Ideally add a couple of perf test in ``vb_suite/gil.py``. *sometimes* releasing the kill is a slight negative (e.g. if the op itself is extremely fast, e.g. < 1ms).